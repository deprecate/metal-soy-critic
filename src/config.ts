import * as fs from 'fs';
import * as path from 'path';
import * as process from 'process';

const CONFIG_FILE_NAME = '.soycriticrc';

export interface ImplicitParamsMap {
  [nameOrRegex: string]: string | Array<string>
}

export interface Config {
  callToImportRegex: string
  callToImportReplace: string
  implicitParams: ImplicitParamsMap
}

export const DEFAULT_CONFIG: Config = {
  callToImportRegex: '(.*)',
  callToImportReplace: '{$1}',
  implicitParams: {}
};

export function readConfig(): Config {
  const filePath = getConfigFilePath();
  let config = {};

  if (filePath) {
    try {
      const buffer = fs.readFileSync(filePath);

      config = JSON.parse(buffer.toString('utf8'));
    } catch(e) {}
  }

  return {...DEFAULT_CONFIG, ...config};
}

export function getConfigFilePath(): string | null {
  let currentPath = process.cwd();

  while (currentPath !== '/') {
    const nextPath = path.join(currentPath, '/', CONFIG_FILE_NAME);

    if (fs.existsSync(nextPath)) {
      return nextPath;
    }
    currentPath = path.dirname(currentPath);
  }

  return null;
}
