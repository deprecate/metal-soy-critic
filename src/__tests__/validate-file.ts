import {getFixturePath}  from '../../test/util';
import validateFile from '../validate-file';
import {DEFAULT_CONFIG} from '../config';

describe('Validate file', () => {
  function validate(fileName, config = DEFAULT_CONFIG) {
    return validateFile(getFixturePath(fileName), config);
  }

  test('should pass', () => validate('Test.soy').then(result => {
    expect(result.status).toBe(true);
  }));

  test('should pass; method param', () => validate('MethodParam.soy').then(result => {
    expect(result.status).toBe(true);
    expect(result.messages.length).toBe(0);
  }));

  test('should pass; valid import', () => validate('CaseImport.soy').then(result => {
    expect(result.status).toBe(true);
    expect(result.messages.length).toBe(0);
  }));

  test('should fail; missing params', () => validate('MissingParams.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; missing internal', () => validate('MissingInternal.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; missing import', () => validate('MissingCallImport.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; missing required', () => validate('MissingRequired.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; invalid namespace', () => validate('InvalidNamespace.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; missing render', () => validate('MissingRender.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; invalid javascript', () => validate('InvalidJS.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; noop events', () => validate('NoopEvents.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; sorted params', () => validate('SortedParams.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; sorted param declarations', () => validate('SortedParamDeclarations.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; sorted map keys', () => validate('UnsortedMap.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; sorted map keys', () => validate('BadTernaryElvis.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; bad whitespace', () => validate('BadWhitespace.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));

  test('should fail; doc params deprecated', () => validate('DocParams.soy').then(result => {
    expect(result.status).toBe(false);
    expect(result.messages).toMatchSnapshot();
  }));
});
