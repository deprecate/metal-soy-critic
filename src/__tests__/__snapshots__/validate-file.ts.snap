// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate file should fail; bad whitespace 1`] = `
Array [
  "There seems to be extra [33mwhitespace[39m on these lines:

[31mLine 8: {@param title: string[7m }[27m[39m",
]
`;

exports[`Validate file should fail; doc params deprecated 1`] = `
Array [
  "Param declarations in docs is [33mdeprecated[39m, see these templates:

[31mrender[39m
[31motherHelper[39m",
]
`;

exports[`Validate file should fail; invalid javascript 1`] = `
Array [
  "Failed to parse component (javascript) file",
]
`;

exports[`Validate file should fail; invalid namespace 1`] = `
Array [
  "Parsing failed at 1:1. Expecting:
'{namespace'",
]
`;

exports[`Validate file should fail; missing import 1`] = `
Array [
  "It looks like the following component calls are missing an [33mimport[39m:

[31mMyExternalComponent[39m",
]
`;

exports[`Validate file should fail; missing internal 1`] = `
Array [
  "Based on their name, these attributes should have the [33m.internal()[39m config added:

[31mcount_[39m",
]
`;

exports[`Validate file should fail; missing params 1`] = `
Array [
  "These params can't be found in [33mSTATE[39m or your [33mClass[39m:

[31moptionalInfo[39m
[31moptionalInfo2[39m
[31moptionalInfo3[39m",
]
`;

exports[`Validate file should fail; missing render 1`] = `
Array [
  "Cannot find a [33m.render[39m template, do you have a typo?

[31mrander[39m",
]
`;

exports[`Validate file should fail; missing required 1`] = `
Array [
  "These attributes are [33mrequired[39m in your Soy Template but not in your Component:

[31mtitle[39m",
  "These attributes are [33m.required()[39m in your Component but not in your Template:

[31mdescription[39m",
]
`;

exports[`Validate file should fail; noop events 1`] = `
Array [
  "These calls are not components, so their [33mevents[39m param will do nothing:

[31mOtherComponent.Helper[39m",
]
`;

exports[`Validate file should fail; sorted map keys 1`] = `
Array [
  "These [33mmap keys[39m should be [33msorted[39m:

[31mLines 10 to 12[39m
[31mLines 18 to 20[39m",
]
`;

exports[`Validate file should fail; sorted map keys 2`] = `
Array [
  "Use the [33mElvis (?:)[39m operator instead of a ternary for [33mdefault[39m values:

[31m$optionalInfo - Line 20[39m",
]
`;

exports[`Validate file should fail; sorted param declarations 1`] = `
Array [
  "These templates need their [33mparam[39m declarations [33msorted[39m:

[31m.render[39m",
]
`;

exports[`Validate file should fail; sorted params 1`] = `
Array [
  "Please [33msort[39m the params in these [33mcalls[39m:

[31mOtherComponent.render - Lines 8 to 11[39m",
]
`;

exports[`Validate file should pass; missing params 1`] = `Array []`;
